libnd4j/include/ops/declarable/helpers/cpu/split.cpp

```diff
@@ -18,26 +18,111 @@
+                        memcpy(outArrs[i]->bufferAsT<T>(), x, memAmountToCopy * sizeofT);
+                        x += memAmountToCopy;
+                    }
+                    return;
+                }
+
+                const bool isXcontin = input.strideAt(axis) == 1 && input.ordering() == 'c';
+                bool areOutsContin = true;
+                bool allSameOrder = true;
+
+                if (isXcontin) {
+                    for (uint i = 0; i < numSplits; ++i) {
+                        areOutsContin &= outArrs[i]->strideAt(axis) == 1;
+                        allSameOrder &= outArrs[i]->ordering() == input.ordering();
+                        if (!areOutsContin || !allSameOrder)
+                            break;
+                    }
+                }
+
+                const bool luckCase2 = isXcontin && areOutsContin && allSameOrder;
+
+                if (luckCase2) {
+
+                    const uint xDim = input.sizeAt(axis);
+
+                    for (uint i = 0; i < input.lengthOf() / xDim; ++i) {
+
+                        T* x = xBuff + xDim * i;
+
+                        for (uint j = 0; j < numSplits; ++j) {
+                            const auto zDim = outArrs[j]->sizeAt(axis);
+                            T* z = outArrs[j]->bufferAsT<T>() + zDim * i;
+                            memcpy(z, x, zDim * sizeofT);
+                            z += zDim;
+                            x += zDim;
+                        }
+                    }
+
+                    return;
+                }
+
+                uint zDim = outArrs[0]->sizeAt(axis);
+                // general case
+
+                auto func = PRAGMA_THREADS_FOR{
+
+                    Nd4jLong coords[MAX_RANK];
+                    for (auto i = start; i < stop; i += increment) {
+
+                        shape::index2coords(i, input.getShapeInfo(), coords);
+                        const auto xOffset = shape::getOffset(input.getShapeInfo(), coords);
+
+                        uint outArrIdx = 0;
+
+                        while (coords[axis] >= zDim) {
+                            coords[axis] -= zDim;
+                            ++outArrIdx;
+                        }
+
+                        T* z = outArrs[outArrIdx]->bufferAsT<T>();
+                        const auto zOffset = shape::getOffset(outArrs[outArrIdx]->getShapeInfo(), coords);
+                        z[zOffset] = xBuff[xOffset];
+                    }
+                };
+
+                samediff::Threads::parallel_for(func, 0, input.lengthOf());
+            }
+
+            void split(nd4j::LaunchContext* context, const NDArray& input, std::vector<NDArray*>& outArrs, const int axis) {
+                BUILD_SINGLE_SELECTOR(input.dataType(), split_, (input, outArrs, axis), LIBND4J_TYPES);
+            }
+
+            BUILD_SINGLE_TEMPLATE(template void split_, (const NDArray& input, const std::vector<NDArray*>& outArrs, const int axis), LIBND4J_TYPES);
```

Comment:
JFYI: if you use BUILD_X_SELECTOR on the method declared in the same compilation unit, you don't need the BUILD_X_TEMPLATE  then, because all templates will be instantiated by selector itself.

Reply:
removed BUILD_SINGLE_TEMPLATE per request as it is redundant